notificationworkerlambda/src/main/scala/com/gu/notifications/worker/delivery/fcm/FcmClient.scala

```diff
@@ -14,12 +14,12 @@ import com.gu.notifications.worker.utils.UnwrappingExecutionException
 
 import java.io.ByteArrayInputStream
 import java.util.UUID
-import java.util.concurrent.Executor
-import scala.concurrent.{ExecutionContextExecutor, Future, Promise}
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
 import scala.util.control.NonFatal
 import scala.util.{Failure, Success, Try}
 
-class FcmClient private (firebaseMessaging: FirebaseMessaging, firebaseApp: FirebaseApp, config: FcmConfig)
+
+class FcmClient (firebaseMessaging: FirebaseMessaging, firebaseApp: FirebaseApp, config: FcmConfig)
```

Comment:
Are we sure it's ok to make this class public?

Reply:
Yeah, a great question. I wondered about this too and here's my current thinking/understanding:
- The `FcmClient` object exposes a public method which creates a success/failure of a new instance of the `FcmClient`
- The public apply method also handles creating the connection to firebase and initialising our "firebase app", which is a prerequisite to being able to instantiate a new `FcmClient`
- This public method is only used in the AndroidSender when we first start the lambda: https://github.com/guardian/mobile-n10n/blob/main/notificationworkerlambda/src/main/scala/com/gu/notifications/worker/AndroidSender.scala#L23

To me the flow looks to have been implemented this way in order to:
- encapsulate the option-setting, app-initialisation and class instantiation all inside a `Try`
- provide a guide to users of the `FcmClient` about how to use the class

I think for our purposes, making the class public:
- allows us to test more easily (fewer mocks and simpler tests)
- doesn't impact how the AndroidSender currently interfaces with the client object and class

If in the future we needed to re-write the android sender lambda we would always need to instantiate with the given options, the current mechanism (using the public object's apply method) is a neat abstraction and making the class public doesn't stop us from continuing to use this.

Sorry, a bit of a ramble - what do you think? Maybe there are other concerns/considerations I've accidentally overlooked!

Reply:
I think thats a great explanation thank you for going into so much detail. 👍

Reply:
I've also tried to use scalatest's PrivateMethodTester but failed spectacularly, possibly because I need to invoke the private apply method and then use the resulting instance to invoke sendNotification...